### PR TITLE
Add a documentation page for lock in FW

### LINKED ISSUE #12561: [Lock] Add missing lock connection string in FrameworkExtension
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#34043
| PR author(s) | @jderusse
| Merged in    | 4.4

### PR DESCRIPTION
**Pull request by jderusse**
This PR add a dedicated page to explain how to use Lock with FrameworkBundle

This PR slightly depends on https://github.com/symfony/symfony/pull/34043 (PDO and Zookeeper DSN)

I should have pushed this PR years ago :cry:  sorry about that...

Fixes: #12561

### CODE REVIEW

File: reference/configuration/framework.rst [hunk 1]
```diff
@@ -2632,20 +2632,7 @@ A list of lock stores to be created by the framework extension.
         # config/packages/lock.yaml
         framework:
             # these are all the supported lock stores
```
Comment (OskarStark) [inline reference/configuration/framework.rst]: Please remove this line

File: lock.rst [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+            'lock' => '%env(LOCK_DSN)%',
+
+            // named locks
+            'lock' => [
+                'invoice' => ['semaphore', 'redis://r2.docker'],
+                'report' => 'semaphore',
+            ],
+        ]);
+
+Locking a resource
```
Comment (OskarStark) [inline lock.rst]: Resource

File: lock.rst [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    // src/Controller/PdfController.php
+    namespace App\Controller;
+
+    use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+    use Symfony\Component\Lock\LockInterface;
+
+    class PdfController extends AbstractController
+    {
+        /**
+         * @Route("/download/term-of-use.pdf")
```
Comment (OskarStark) [inline lock.rst]: termS

File: lock.rst [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        }
+    }
+
+.. caution::
+
+    The same instance of ``LockInterface`` won't block when calling `acquire``
+    multiple times. If several services share the same ``lock`` service, they
+    won't lock each other, use ``LockFactory`` instead.
+
+Locking a dynamic resource
```
Comment (OskarStark) [inline lock.rst]: ... Dynamic Resource

File: lock.rst [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    // src/Controller/PdfController.php
+    namespace App\Controller;
+
+    use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+    use Symfony\Component\Lock\LockInterface;
+
+    class PdfController extends AbstractController
+    {
+        /**
+         * @Route("/download/${version}/term-of-use.pdf")
```
Comment (OskarStark) [inline lock.rst]: The $ looks wrong here

Comment (OskarStark) [reply lock.rst]: termS

File: lock.rst [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+            // heavy computation
+            $myPdf = $pdf->getOrCreatePdf($version);
+
+            $lock->release();
+
+            // ...
+        }
+    }
+
+Named lock
```
Comment (OskarStark) [inline lock.rst]: Lock

File: lock.rst [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+ type-hinting it with :class:`Symfony\\Component\\Lock\\LockInterface`.
+
+Symfony also provide a corresponding factory and store following the same rules
+(e.g. ``invoice`` generates a ``lock.invoice.factory`` and
+``lock.invoice.store``, both can be injected automatically by naming
+respectively ``$invoiceLockFactory`` and ``invoiceLockStore`` and type-hinted
+with :class:`Symfony\\Component\\Lock\\LockFactory` and
+:class:`Symfony\\Component\\Lock\\PersistingStoreInterface`)
+
+Blocking store
```
Comment (OskarStark) [inline lock.rst]: Store

File: lock.rst [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+--------------
+
+If you want to use the `RetryTillSaveStore` for :ref:`non-blocking locks <lock-blocking-locks>`,
```
Comment (OskarStark) [inline lock.rst]: Please use. 2 ticks around RetryTillSaveStore

**Review by OskarStark — CHANGES_REQUESTED** [review]

**Comment by jderusse** [comment]
RFR

File: lock.rst [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+If you want to use the ``RetryTillSaveStore`` for :ref:`non-blocking locks <lock-blocking-locks>`,
+you can do it by :doc:`decorating the store </service_container/service_decoration>` service:
+
+.. code-block:: yaml
+
+    lock.default.retry_till_save.store:
+        class: Symfony\Component\Lock\Store\RetryTillSaveStore
+        decorates: lock.default.store
+        arguments: ['@lock.default.retry.till.save.store.inner', 100, 50]
```
Comment (clemherreman) [inline lock.rst]: I think there is a typo in the inner service name

```suggestion
        arguments: ['@lock.default.retry_till_save.store.inner', 100, 50]
```

**Review by clemherreman — COMMENTED** [review]
If I were a real reviewer, I'd say this looks good to merge :) Thank you @jderusse

File: lock.rst (hunk 9 above)
Comment (jderusse) [reply lock.rst]: indeed. good catch

File: lock.rst [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+Each name becomes a service where the service id suffixed by the name of the
+lock (e.g. ``lock.invoice``). An autowiring alias is also created for each lock
+using the camel case version of its name suffixed by ``Lock`` - e.g. ``invoice``
+can be injected automatically by naming the argument ``$invoiceLock`` and
+type-hinting it with :class:`Symfony\\Component\\Lock\\LockInterface`.
+
+Symfony also provide a corresponding factory and store following the same rules
+(e.g. ``invoice`` generates a ``lock.invoice.factory`` and
+``lock.invoice.store``, both can be injected automatically by naming
+respectively ``$invoiceLockFactory`` and ``invoiceLockStore`` and type-hinted
```
Comment (maxhelias) [inline lock.rst:276]: You miss "$" no ? $invoiceLockStore ?

**Review by maxhelias — COMMENTED** [review]
Thanks for the reminder, i forgot some notion 😁

File: lock.rst [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+    // critical section of code
+    $service->method();
+
+    $lock->release();
+
+Installation
+------------
+
+In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
+install messenger:
```
Comment (wouterj) [inline lock.rst:29]:
```suggestion
install the Lock component:
```

File: lock.rst [hunk 12]
```diff
@@ -0,0 +1,291 @@
+.. index::
+    single: Lock
+
+Dealing with concurrency with Lock
```
Comment (wouterj) [inline lock.rst:4]:
```suggestion
Dealing with concurrency with Locks
```

I find "with Lock" a bit weird, as you actually mean "with the Lock component". Looking at the titles of all other guides, I think we should either call this article "Locks" or my suggestion (to make a normal sentence).

File: lock.rst [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+==================================
+
+When a program runs concurrently, some part of code which modify shared
+resources should not be accessed by multiple processes at the same time.
+Symfony's :doc:`Lock </components/lock>` provides a locking mechanism to ensure
```
Comment (wouterj) [inline lock.rst:9]:
```suggestion
Symfony's :doc:`Lock component </components/lock>` provides a locking mechanism to ensure
```

File: lock.rst [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+
+            $lock->release();
+
+            // ...
+        }
+    }
+
+.. caution::
+
+    The same instance of ``LockInterface`` won't block when calling `acquire``
```
Comment (wouterj) [inline lock.rst:183]:
```suggestion
    The same instance of ``LockInterface`` won't block when calling ``acquire``
```

File: lock.rst [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+    multiple times. If several services share the same ``lock`` service, they
+    won't lock each other, use ``LockFactory`` instead.
```
Comment (wouterj) [inline lock.rst:185]: I don't really understand this caution. The above example, when run concurrently *will* block eachother, right?

**Review by wouterj — APPROVED** [review]
Sorry for completely missing this PR @jderusse! I think it's a very great step (we're trying to move all important information from the component to the main guides in the docs).

I've left a few comments, most of them are minor and can be fixed while merging. So don't worry if you don't have time to fix them now :) Imho, it's ready for merge

File: lock.rst (hunk 10 above)
Comment (jderusse) [reply lock.rst:276]: good catch thanks

File: lock.rst (hunk 15 above)
Comment (jderusse) [reply lock.rst:185]: Depends, what you call `concurrently`, when running currently in the same process (think react-php), the same instance of Lock (think shared service) wont block.

Bellow case where lock does not apply
```php
$lock = new Lock();
$lock->acquire(); 
$lock->acquire(); // no lock (same instance)
```

```php
$lock = new Lock();
$a = new ReactControler($lock); 
$b = new ReactControler($lock); // Both $a and  share same lock instance and wont block each other
```

Comment (jderusse) [reply lock.rst:185]: I would appreciate help to rephrase my sentence :)

Comment (jderusse) [reply lock.rst]: reworded

File: lock.rst [hunk 16]
```diff
@@ -0,0 +1,292 @@
+.. index::
+    single: Lock
+
+Dealing with concurrency with Locks
```
Comment (OskarStark) [inline lock.rst:4]:
```suggestion
Dealing with Concurrency with Locks
```

Comment (jderusse) [reply lock.rst:4]: why an upper-case?

note: we also use `concurrency` in lower case in `http_client.rst` 

> That's the trick when concurrency is desired

Comment (wouterj) [reply lock.rst:4]: The title-case standard we're using says "Every word should be capitialized, except from closed-class words" (https://symfony.com/doc/current/contributing/documentation/standards#english-language-standards). "concurrency" is not a closed class word (but "with" is)

Comment (jderusse) [reply lock.rst]: thanks for the reference :+1:

Comment (wouterj) [reply lock.rst]: Thanks for the quick actions :)

**Review by maxhelias — APPROVED** [review]

**Comment by wouterj** [comment]
Thanks a lot Jérémy! This PR has been forgotten about for to many times, my apologies. It's now merged into the 4.4 version of the docs.

File: lock.rst (hunk 15 above)
Comment (wouterj) [reply lock.rst]: I've revisited this caution in 1170a7c59 . Can you please check if the new contents of this caution are correct?

Comment (jderusse) [reply lock.rst]: sounds good.
Thanks for your help

**Pull request merged**